core/src/main/java/org/apache/iceberg/BaseTableScan.java

```diff
@@ -167,59 +154,52 @@ public TableScan asOfTime(long timestampMillis) {
   @Override
   public TableScan option(String property, String value) {
     ImmutableMap.Builder<String, String> builder = ImmutableMap.builder();
-    builder.putAll(options);
+    builder.putAll(context.options());
     builder.put(property, value);
 
     return newRefinedScan(
-        ops, table, snapshotId, schema, rowFilter, ignoreResiduals,
-        caseSensitive, colStats, selectedColumns, builder.build());
+        ops, table, schema, context.options(builder.build()));
```

Comment:
I think it would be better to add `withOption(String property, String value)` to context and move the map builder there, instead of replacing all options. That keeps this simple and delegates work to the class with the most context (tell, don't ask).